### PR TITLE
Convert duplicated code into helper functions

### LINKED ISSUE #44: Code duplication as a result of splitting `chain_sim()`
The change [here](https://github.com/epiverse-trace/epichains/pull/33) that I'm least sure about is the splitting of `chain_sim`. While cyclomatic complexity and complexity of function arguments have been reduced (which is great) this has come at the expense of some code duplication (which is bad). I don't have an immediate satisfactory solution but this would be good to think about further at some point.

_Originally posted by @sbfnk in https://github.com/epiverse-trace/epichains/pull/33#pullrequestreview-1600869715_

### PR DESCRIPTION
**Pull request by jamesmbaazam**
* **Please check if the PR fulfils these requirements**
  
- [x] I have read the CONTRIBUTING guidelines 
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

- It refactors `simulate_chains()`, `simulate_summary()`, and `likelihood()` to remove duplicated code by converting them into internal helper functions follows: 

	- Simulation related operations have been modularised into `.init_susc_pop()`, `.sample_possible_offspring()`, and 
	`.get_susceptible_offspring()`. 
	- Input checking is now remitted to 3 functions, `.check_sim_args()`, which checks the shared arguments of the `simulate_*		()` functions, and 2 smaller functions, `.check_statistic_args()` which checks `stat_max` and `statistic` together in 	multiple functions, and `.check_time_args()`, which checks the time-related arguments: `generation_time`, `t0`, and `tf`.

- Tests and documentation, where necessary, have been added.

* **What is the current behaviour?** (You can also link to an open issue here)

This is a refactoring exercise and leads to no user-facing behaviour change.

* **What is the new behaviour (if this is a feature change)?**
  
  This is a refactoring exercise and leads to no user-facing behaviour change.
  
* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No.

* **Other information**:

- This PR closes #44.

### CODE REVIEW

File: R/simulate.r [hunk 1]
```diff
@@ -139,40 +139,27 @@ simulate_chains <- function(index_cases,
...
-  }
-  checkmate::assert_numeric(
-    t0,
-    lower = 0, finite = TRUE
-  )
-  checkmate::assert_number(
-    tf,
-    lower = 0
+  # Check inputs
+  func_name <- as.character(match.call()[[1]])
```
Comment (sbfnk) [inline R/simulate.r:143]: is there any benefit of doing this over just saying `func_name = "simulate_chains"` below which would be simpler?

File: R/simulate.r [hunk 2]
```diff
@@ -139,40 +139,27 @@ simulate_chains <- function(index_cases,
...
+  # We need to check if the generation time and tf are specified here before
+  # calling .check_sim_args() to not obfuscate the meaning of missing()
+  # in the checker function.
+  generation_time_specified <- !missing(generation_time)
```
Comment (sbfnk) [inline R/simulate.r:147]: could this not be solved in a simpler way by replacing
```r
    if (generation_time_specified) {
```
with
```r
    if (!is.null(generation_time)) {
```
below?

File: R/simulate.r [hunk 3]
```diff
@@ -206,30 +193,20 @@ simulate_chains <- function(index_cases,
...
-      c(
-        list(n = sum(n_offspring[sim])),
-        pars
+    # simulate the next possible offspring
+    next_gen <- .sample_possible_offspring(
+      offspring_func = offspring_dist,
+      offspring_func_pars = pars,
+      n_offspring = n_offspring,
+      chains = sim
       )
```
Comment (sbfnk) [inline R/simulate.r:202]:
```suggestion
    )
```

File: R/simulate.r [hunk 4]
```diff
@@ -206,30 +193,20 @@ simulate_chains <- function(index_cases,
...
+    # from all possible offspring, get those that could be infected
+    next_gen <- .get_infectible_offspring(
```
Comment (sbfnk) [inline R/simulate.r:204]:
```suggestion
    next_gen <- .get_susceptible_offspring(
```
to be more in line with the rest of the terminology

File: R/helpers.R [hunk 5]
```diff
@@ -35,6 +35,91 @@
...
+#' This function is used internally, and input checking is not
+#' performed here, only in the context where it is used. Using it directly
+#' is not recommended.
+#' @inheritParams simulate_chains
+#' @param new_offspring A vector of the possible new offspring per chain
+#' produced by [.sample_possible_offspring()]
+#' @return A vector of the number of offspring that can be infected given the
+#' current susceptible population size
+#' @keywords internal
+.get_infectible_offspring <- function(new_offspring,
```
Comment (sbfnk) [inline R/helpers.R:103]:
```suggestion
.get_susceptible_offspring <- function(new_offspring,
```

File: R/helpers.R [hunk 6]
```diff
@@ -35,6 +35,91 @@
...
+                                      susc_pop,
+                                      pop) {
+  # We first adjust for the case where susceptible can be Inf but prob can only
+  # be maximum 1.
+  binom_prob <- min(1, susc_pop / pop, na.rm = TRUE)
+  # Sample the number of infectible offspring from all possible offspring
+  infectible_offspring <- stats::rbinom(
```
Comment (sbfnk) [inline R/helpers.R:110]:
```suggestion
  susceptible_offspring <- stats::rbinom(
```

File: R/helpers.R [hunk 7]
```diff
@@ -35,6 +35,91 @@
...
+    n = length(new_offspring),
+    size = new_offspring,
+    prob = binom_prob
+  )
+  return(infectible_offspring)
```
Comment (sbfnk) [inline R/helpers.R:115]:
```suggestion
  return(susceptible_offspring)
```

File: R/helpers.R [hunk 8]
```diff
@@ -35,6 +35,91 @@
...
+  # check that offspring distribution returns integers
+  stopifnot(
+    "Offspring distribution must return integers" =
+      !all(possible_new_offspring %% 1 > 0)
+  )
+
+  return(possible_new_offspring)
+}
+
+#' Sample the number of infectible offspring from all possible offspring
```
Comment (sbfnk) [inline R/helpers.R:90]:
```suggestion
#' Sample the number of susceptible offspring from all possible offspring
```

File: inst/WORDLIST [hunk 9]
```diff
@@ -30,6 +30,7 @@ gborel
 geq
 infectee
 infectees
+infectible
```
Comment (sbfnk) [inline inst/WORDLIST:33]: can go

File: tests/testthat/test-checks.R [hunk 10]
```diff
@@ -15,4 +15,165 @@ test_that("Checks work", {
     .check_generation_time_valid(function(x) 3),
     "Must have length"
   )
+  expect_error(
```
Comment (sbfnk) [inline tests/testthat/test-checks.R:18]: maybe add some comments here for code reviewers to see what is being checked? It's a bit hard to see with so many arguments. Alternatively define a default function to avoid having to repeat the same arguments over and over again. See here for an example of this approach:
https://github.com/epiforecasts/EpiNow2/blob/df1fdc8bcb361da308037fbb9c8deb6f3fc45235/tests/testthat/test-estimate_infections.R#L7

**Review by sbfnk — CHANGES_REQUESTED** [review]
Great stuff. Also

```r
> cyclocomp::cyclocomp(simulate_chains )
[1] 15
```

so can switch the lintr on again.

File: R/simulate.r (hunk 1 above)
Comment (jamesmbaazam) [reply R/simulate.r:143]: It's just a guard against potential name changes.

File: tests/testthat/test-checks.R (hunk 10 above)
Comment (jamesmbaazam) [reply tests/testthat/test-checks.R]: Thanks for pointing me to this. I've restructured the tests here 31d09158e831a52233932b3652ad2fbc2c0de33b.

File: R/simulate.r (hunk 2 above)
Comment (jamesmbaazam) [reply R/simulate.r]: Good point. I've made the change here 53c92977d0f43f1b04868531ceabb8253dcd8abf.

File: inst/WORDLIST (hunk 9 above)
Comment (jamesmbaazam) [reply inst/WORDLIST]: Thanks. I've updated the WORDLIST.

File: tests/testthat/test-checks.R (hunk 10 above)
Comment (sbfnk) [reply tests/testthat/test-checks.R]: Awesome, this is much clearer.

File: R/checks.R [hunk 11]
```diff
@@ -18,3 +18,91 @@
...
+    }
+    checkmate::assert_number(
+      tf,
+      lower = 0
+    )
+    checkmate::assert_numeric(
+      t0,
+      lower = 0, finite = TRUE
+    )
+  }
```
Comment (sbfnk) [inline R/checks.R:78]: Sorry I didn't think of this earlier but: If we made this snippet a separate function `.check_time_args` then we could do away with passing the function name (which I found really hard to parse, but that may of course be just me). What do you think?

**Review by sbfnk — APPROVED** [review]
All good. Just one last comment.

File: R/checks.R (hunk 11 above)
Comment (jamesmbaazam) [reply R/checks.R:78]: Ah, yes! That's a good idea. Thanks. I'm working on that now.

Comment (jamesmbaazam) [reply R/checks.R]: I've now made the change here 8bd77e8aac80fba17786a0ecc11f83483aac57df.
